Introduce image build / push commands

iocanel opened the pull request:
This pull request introduces the following:

1. The ability to use force extensions via System properties.
2. CLI commands for building container images.

The commands work like:

```
quarkus image build 
quarkus image push
```

for build commands user has access to more specific sub-commands which expose container image extension specific options:

```
quarkus image build docker 
quarkus image build jib
quarkus image build openshift
```

The commands above also force the extension, so it doesn't need to be in the pom.

This is still work in progress as there are some tasks that will most probably require collaboration:
- [ ] Need to find a cleaner way for forcing extensions.
- [ ] Need to discuss if the proposed format is ideal.
- [ ] Need to align push with build commands

iocanel committed:
feat: introduce image build / push commands

iocanel commented:
@aloubyansky: I am looking for a cleaner way to force the extensions. Do you have any ideas / pointer?

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/QuarkusBootstrapProvider.java:
```diff
@@ -205,14 +225,15 @@ protected CuratedApplication doBootstrap(QuarkusBootstrapMojo mojo, LaunchMode m
 
             QuarkusBootstrap.Builder builder = QuarkusBootstrap.builder()
                     .setAppArtifact(appModel.getAppArtifact())
+                    .setForcedDependencies(forcedExtensions)
```
Oh, the current impl is actually broken. The thing is `forcedDependencies(...)` is actually called a few lines overriding what you are setting but it doesn't actually matter because the model is resolved a few lines above.

aloubyansky commented:
TBH, I would prefer having a `quarkus:image` goal in Maven and `quarkusImage` task in Gradle and the CLI command calling those. This would enhance those tools as well making users who prefer those over the CLI happy and avoid introducing this new config option, which is looking questionable to me.

iocanel commented:
> TBH, I would prefer having a `quarkus:image` goal in Maven and `quarkusImage` task in Gradle and the CLI command calling those. This would enhance those tools as well making users who prefer those over the CLI happy and avoid introducing this new config option, which is looking questionable to me.

That's not a bad idea. I can start adding those goals/tasks ASAP.

What do you mean by `avoid introducing this new config option`? 

What do you think about the approach I am following to detect the quarkus version? Is there a cleaner more reliable way to do so?

aloubyansky commented:
I meant `quarkus.application.forced-extensions`.

aloubyansky commented:
> What do you think about the approach I am following to detect the quarkus version? Is there a cleaner more reliable way to do so?

No, that's not good, actually. There are `io.quarkus` artifacts that aren't part of the quarkus repo. Looking for `quarkus-core` artifactId and `io.quarkus` groupId would be more reliable.
However, a better approach would be to look for versions of the extensions are forcing among the dependency constraints (imported BOMs), assuming that these extensions are managed by the BOMs (which they are in your case).
In a Maven mojo you could get the constraints from the `MavenProject` with `getDependencyManagment()` or possibly `getManagedVersionMap()` (haven't used this one, I think).
In Gradle it's a bit more involved. I'll look for an example for you.

Also you may want to check whether those extensions are already present among the app dependencies before you enforce them, although it's unlikely to do anything bad if you don't.

iocanel commented:
> I meant `quarkus.application.forced-extensions`.

It's not clear to me how we would avoid using this option by delegating to a maven goal / gradle task, but maybe I miss something.

aloubyansky commented:
If you add specialized goals similar to what you did in the cli, e.g. `quarkus:jib`, etc, they would know what to enforce and they would call `forcedDependencies(...)` directly.
So the CLI `jib` command would simply call the `quarkus:jib` goal that would do all the work.

aloubyansky commented:
Here is another reason why relying `io.quarkus:quarkus-core` version isn't a good idea. https://maven.repository.redhat.com/ga/com/redhat/quarkus/platform/quarkus-bom/2.2.3.SP2-redhat-00001/quarkus-bom-2.2.3.SP2-redhat-00001.pom
````
<dependency>
<groupId>io.quarkus</groupId>
<artifactId>quarkus-container-image-jib</artifactId>
<version>2.2.3.Final</version>
</dependency>
<dependency>
<groupId>io.quarkus</groupId>
<artifactId>quarkus-container-image-openshift</artifactId>
<version>2.2.3.Final-redhat-00013</version>
</dependency>
````

aloubyansky commented:
> > I meant `quarkus.application.forced-extensions`.
> 
> It's not clear to me how we would avoid using this option by delegating to a maven goal / gradle task, but maybe I miss something.

Basically, imagine we don't have the CLI and you need to implement this directly in the Maven and Gradle plugins. We wouldn't need `quarkus.application.forced-extensions` there, right?

aloubyansky commented:
In Maven, I suppose, you would extend the `BuildMojo` and override its `protected List<Dependency> forcedDependencies(LaunchMode mode)` inherited from the `QuarkusBootstrapMojo`

aloubyansky commented:
@iocanel I could help with the version resolutions in Maven/Gradle once you have the basic classes in place. It looks like quite an effort overall.

iocanel committed:
refactor: make BuildMojo extendable

iocanel committed:
feat: introduce image-build and image-push goals

iocanel committed:
feat: add imageBuild and imagePush gradle tasks

iocanel commented:
@aloubyansky: Added some maven mojos and gradle tasks. 
It is still work in progress, however the maven bits work. The gradle stuff fail to force dependencies and this is where I'd like your aid. Could you provide some pointers?

aloubyansky commented:
The Maven part looks good. We need to do a bit of refactoring in the Gradle plugin to enable forced dependencies. I'll look into that.

iocanel commented:
> The Maven part looks good. We need to do a bit of refactoring in the Gradle plugin to enable forced dependencies. I'll look into that.

Cool, let me know what that is done. Meanwhile, I'll try to improve the commands to better integrate with the mojos/tasks.

iocanel commented:
@aloubyansky: I tried to continue my work on the gradle commands, but it seems that `forcedDependencies` still don't work.
With a quick glimpse at the `AppModelGradleResolver` it seems that resolving the artifact with a collection of dependencies is not implemented.

aloubyansky commented:
Here is how the classpath part could be implemented https://github.com/aloubyansky/quarkus/commit/925717595cc96dc27cc036d36d528a5b470c4cc0
There are still a few problems with it. E.g. the task isn't declaring any output, which isn't adhering to the Gradle best practices (supporting incremental builds). The version resolution (normalizing the coords) isn't great, although for your specific case, could be good enough.
The other part if the config. I was wondering whether the task could create a new config source that would contain the extra config options for the build and that could also be registered the task's output.

@glefloch this could be of interest to you. I think you may have some suggestions for us.

glefloch commented:
I really like the idea! From what I understand, `imageBuild`/`imagePush` tasks can be seen as `pre-configuration` task for the `quarkusBuild` task. They provides both properties and forced extensions.

In gradle it's not a good practice to try to configure a task from another task(e.g loading the `quarkusBuild` task instance   from `imageBuild` task and setting properties). This will be prevented in the future as they are pushing for task cache introduction. 

Another issue we may face is that running `./gradlew quarkusBuild imageBuild imagePush` should run `quarkusBuild` only once with a combination of all properties / extension. And as mention by @aloubyansky, running it again should result in a `UP-TO-DATE` status. Thus we cannot base `imageBuild` / `imagePush` task on the `QuarkusBuild.class` class.

I think I would implement that mostly in the same way Alexey did, meaning: 
* Creating a `QuarkusBuildConfigurationProvider` (similar to `QuarkusTaskWithForcedDependencies`) class 
* This class would have one input containing the builder property such as: 

````
public class QuarkusBuildConfigurationProvider extends QuarkusTask {
    private Map<String, String> extraProperties;
    private Builder builder;

    private void setExtraProperties(Map<String, String> extraProperties) {
       this.extraProperties = extraProperties;
    }

    @Input
    public void getBuilder() {
        if(builder == null) {
           return Builder.docker;
        }
        return builder;
    }
    @Option(option = "builder", description = "The container image extension to use for building the image (e.g. docker, jib, buildpack, openshift).")
    public void setBuilder(Builder builder) {
        this.builder = builder;
    }

}

````

* In the plugin we can register this class twice with different configurations : 
  ````
   tasks.register("imagePush", QuarkusBuildConfigurationProvider.class, task -> {
      task.finalizedBy("quarkusBuild"); // This may be not needed
      task.setProperties(Map.of("quarkus.container-image.build", "true", "quarkus.container-image.push", "true")
   });
 
   tasks.register("imagePush", QuarkusBuildConfigurationProvider.class, task -> {
      task.finalizedBy("quarkusBuild"); // This may be not needed
      task.setProperties(Map.of("quarkus.container-image.build", "true")
   });
 
  ````

* Then for property propagation,  I think we can either use a file or a `MapProperty`. That will be used as task output. 
* In the `@TaskAction`, we can add the dependency to the correct `deployment` configuration. 

WDYT of this approach @aloubyansky ? 

@iocanel I can work on this if you need help.

aloubyansky commented:
Sounds like it could work.

iocanel commented:
I am already experimenting with this approach. The property propagation part is not 100% clear to me but I am looking into it.

aloubyansky commented:
@glefloch how would you pass the properties to `quarkusBuild`?

aloubyansky commented:
BTW @iocanel working on the refactoring in the Gradle plugin for this to work I realized the limitations we have in Maven. You may have noticed if `generate-code` is executed before, the forced dependencies that are provided by later executed mojos are ignored. It's because we are bootstrapping an app only once for the build. I'll need to look into that and come up with something similar to how it's done in Gradle.

glefloch commented:
@aloubyansky, we can produce a file and consumes it in the `quarkusBuild` task. We could also leverage the Gradle Provider API which can be set as task output/input.

aloubyansky commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusBuild.java:
```diff
@@ -125,14 +144,23 @@ public File getFastJar() {
-                    getProject().getName(), getProject().getVersion().toString()));
+            appModel = extension().getAppModelResolver().resolveManagedModel(
+                    new GACTV(getProject().getGroup().toString(), getProject().getName(), getProject().getVersion().toString()),
+                    forcedDependencies, null, Collections.emptySet());
```
The impl of this method in the Gradle plugin ignores the forced dependencies, which is why it doesn't work. I think I had a branch where I (tried to) implement it. We had some changes in main though and I'll need to rebase it and see if it still works.

aloubyansky commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusBuild.java:
```diff
@@ -149,6 +177,7 @@ public void buildQuarkus() {
                 .setAppArtifact(appModel.getAppArtifact())
                 .setLocalProjectDiscovery(false)
                 .setIsolateDeployment(true)
+                .setForcedDependencies(forcedDependencies)
```
If you are setting an `existingModel` then calling `setForcedDependencies(deps)` here won't have any affect, given that the model has already been resolved.

iocanel replied:
So, this is redundant given that I also pass the forced dependencies in the `resolveManagedModel`. I'll remove it.

iocanel committed:
refactor: cli commands now use mojos and tasks

iocanel commented:
@aloubyansky: As this is currently blocked by the fact that forced dependencies do not seem to work for gradle, I am pondering on how we could proceed with this, since your plate seems to be full.

One alternative would be to document it as a known issue and proceed normally.
An other alternative would be to provide some pointers and where to look.

wdyt?

aloubyansky commented:
@iocanel I agree, unfortunately, forced dependencies aren't on top of my list, atm.

iocanel commented:
@aloubyansky: I think that the PR is ready for review.

geoand reviewed (commented):
I am not terribly thrilled about the fact that we will now need to keep this in sync with the regular container-image code, but nonetheless I do see this as an important feature.

BTW, `QuarkusBuildConfiguration` seems to be an empty file, is it really needed?

iocanel commented:
> I am not terribly thrilled about the fact that we will now need to keep this in sync with the regular container-image code, but nonetheless I do see this as an important feature.

Agree, I'd love to have a way of validating the `cli options -> config` mapping by the compiler. 
However, given that the configuration interface of these extensions is not volatile, I think it's not a blocker. 

> BTW, `QuarkusBuildConfiguration` seems to be an empty file, is it really needed?
Not needed. Will remove it shortly.

aloubyansky commented:
If we don't support these commands in Gradle currently, should we rollback all the Gradle-related changes and throw an error if a CLI command is executed in a Gradle project?
The other thing we need to fix is call `quarkus:generate-code` from the base image command, otherwise it won't work for extensions that generate source code. I am wondering whether you could extend `QuarkusBootstrapMojo` instead of the `BuildMojo` too.

iocanel commented:
> If we don't support these commands in Gradle currently, should we rollback all the Gradle-related changes and throw an error if a CLI command is executed in a Gradle project? 

The gradle commands will work fine, when the extensions are already configured for the project. 
When they are not the commands will fail silently (which does make for a bad user experience).

We could possibly improve the user experience with out of the following:

- Warn users when there are required extensions missing, and provide instructions on how to address.
- Add the required extension(s) as if `quarkus ext add` was called.

> The other thing we need to fix is call `quarkus:generate-code` from the base image command, otherwise it won't work for extensions that generate source code. I am wondering whether you could extend `QuarkusBootstrapMojo` instead of the `BuildMojo` too.

Not sure I get the issue with generate code. I can try the `QuarkusBootstrapMojo` suggestion and let you know.

aloubyansky commented:
If the required extensions aren't present in a Gradle project failing with an error message would be great then.

The image commands assume `compile` has been called, right? In case of Maven users would do `mvn compile quarkus:<image-goal>` to make sure everything was compiled, right? In that case, `quarkus:generate-code` would be covered.

iocanel commented:
> The image commands assume `compile` has been called, right? In case of Maven users would do `mvn compile quarkus:<image-goal>` to make sure everything was compiled, right? In that case, `quarkus:generate-code` would be covered.

The image commands do not assume anything. They just extend the `BuildMojo` which means that they can be used standalone, or combined with any other goal.

aloubyansky commented:
The `BuildMojo` assumes the sources (original and generated) have been compiled though. To use `quarkus:build` reliably would mean to combine it with `compile`, i.e. `mvn compile quarkus:build`. However, our users don't typically execute `quarkus:build` explicitly, they run `mvn package` instead which takes care of everything.
In contrast, the image commands are meant to be called explicitly. The way they are implemented now, the recommended way would be `mvn [clean] compile quarkus:<image-goal>`. The CLI ones could actually add `compile` behind the scenes.

The `DevMojo`, otoh, doesn't assume anything. It checks whether `compile` was explicitly present on the command line and if it wasn't it would invoke `compile` itself.

iocanel committed:
feat: warn gradle user about missing container ext

iocanel commented:
Added warning and detailed instructions for gralde users that are missing a required container image extension.
Maven cli commands now explicitly call compile before the actual image goals.

iocanel commented:
@aloubyansky @geoand: Incorporated suggestion and fixed CI issues, there are still failures which are irrelevant to this PR. Let me know what you think.

geoand commented:
Looks good from my PoV

aloubyansky commented on devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java:
```diff
@@ -116,7 +122,15 @@ public Object buildAll(String modelName, ModelParameter parameter, Project proje
                 project, modelBuilder, appArtifact.getWorkspaceModule().mutable());
         collectExtensionDependencies(project, deploymentConfig, modelBuilder);
-
+        if (mvnDeps != null) {
```
We shouldn't be doing it this way. Can we remove adding these `mvnDeps` to the model?

aloubyansky replied:
These would need to be added to the runtime configuration to make sure all the dependencies are properly resolved and available for the application. This kind of change would need more testing and review.

iocanel replied:
Removed.

iocanel commented:
@aloubyansky Removed the changes I made in my effort to get `forcedDependencies` working for gradle.

aloubyansky commented:
@ebullient would you like to review this one?

ebullient commented on devtools/cli/src/test/java/io/quarkus/cli/image/CliImageMavenTest.java:
```diff
@@ -0,0 +1,68 @@
+
+        // 1 image --help
+        result = CliDriver.execute(project, "image", "--help");
+        Assertions.assertTrue(result.getStdout().contains("Commands:"), "Should list subcommands\n");
```
These are effectively the same tests as in devtools/cli/src/test/java/io/quarkus/cli/CliHelpTest.java. Validation of help output should move there.

What should be here are invocations with dry-run set, so you can verify what parameters were set based on the command line arguments used.

You have a lot of "if you set this argument, these properties should be set", and you should probably verify one or two, especially if there is anywhere "tricky" that future you won't remember to be careful about.

There should be at least one in maven, and one in gradle, that verifies you have the right resulting maven/gradle command (that's what dryrun is for)

iocanel replied:
Tests moved to `CliHelpTest` and `CliImageMavenTest` now uses dry run to check that options are well respected.

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
+
+=== Image building and pushing
+
+The quarkus cli allows building container images without tampering with your project configuration (adding / removing container image extensions).
```
```suggestion
The quarkus CLI allows building container images without tampering with your project configuration (adding / removing container image extensions).
```

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
 ----
+
+
+=== Image building and pushing
```
```suggestion
=== Container images
```

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
+=== Image building and pushing
+
+The quarkus cli allows building container images without tampering with your project configuration (adding / removing container image extensions).
+To perform a build:
```
```suggestion
To build the image for your project:
```

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
+quarkus image build
+----
+
+The build command can be used directly, or a subcommand can be selected. The available subcommands are:
```
```suggestion
The `image build` command can be used directly, or a subcommand can be selected. The available subcommands are:
```

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
+- jib
+- openshift
+
+Each subcommand corresponds to an image building tool supported by quarkus and gives access to specific configuration options.
```
```suggestion
Each subcommand corresponds to an image building tool supported by Quarkus and gives access to specific configuration options.
```

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
+
+Each subcommand corresponds to an image building tool supported by quarkus and gives access to specific configuration options.
+
+For example to use `buildpacks` instead of `docker` and set the builder image of choice:
```
```suggestion
For example, to use a https://buildpacks.io/[Cloud Native Buildpack] with a custom builder image, use the following:
```

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
+
+==== Pushing
+
+Similar to `build` the `push` command works similarly. It allows allows for pushing images and exposes some basic options related to the target registry.
```
```suggestion
The `image push` command is similar to `image build`, and surfaces some basic options required to push images to a target container registry.
```

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
+quarkus image push --registry=<<image registry>> --registry-username=<<registry username>> --registry-passowrd=<<registry password>>
+----
+
+Note: It's not recommended to pass username and password as command line arguments.
```
If it is not recommended to pass the username or password as command line arguments, then it shouldn't be your example. Should the options exist at all? or should the CLI read from a .env file or environment variables or stdin?

Do what is the right thing.

iocanel replied:
Added option to optionally read password from stdin (ala docker login, oc login and so on).
Docs, now use that property.

ebullient commented:
From the perspective of docs changes, I recommend you try the vale linter:
https://github.com/quarkusio/quarkus/blob/main/docs/src/main/asciidoc/doc-contribute-docs-howto.adoc#lint-documentation-changes-with-vale

Specifically, use `-Dvale=git`. It should stick just with the file you've changed. You don't have to fix everything in that file (there is more that won't be you), but it would have caught CLI vs. cli, etc.

iocanel committed:
Update devtools/cli/src/main/java/io/quarkus/cli/Image.java

Co-authored-by: Erin Schnabel <ebullientworks@gmail.com>

iocanel commented:
> From the perspective of docs changes, I recommend you try the vale linter: https://github.com/quarkusio/quarkus/blob/main/docs/src/main/asciidoc/doc-contribute-docs-howto.adoc#lint-documentation-changes-with-vale
> 
> Specifically, use `-Dvale=git`. It should stick just with the file you've changed. You don't have to fix everything in that file (there is more that won't be you), but it would have caught CLI vs. cli, etc.

It seems that you already caught everything there was to cach :smile:

iocanel committed:
feat: allow reading registry password from stdin

aloubyansky commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -556,3 +564,42 @@ alias q=quarkus
+
+=== Container images
+
+The quarkus CLI allows building container images without tampering with your project configuration (adding / removing container image extensions).
```
```suggestion
The Quarkus CLI allows building container images without tampering with your project configuration (adding / removing container image extensions).
```

ebullient replied:
I got the one but not the other?! EEEK. LOL

iocanel committed:
feat: support dry run in image cli, mojo and tasks

iocanel commented:
@ebullient: I think I addressed all your points, can you please have anohter look?

ebullient commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/ImageBuild.java:
```diff
@@ -46,6 +47,9 @@ public List<Dependency> forcedDependencies() {
 
     @Override
     public Map<String, String> forcedProperties() {
+        if (dryRun) {
```
Is there a reason you wouldn't return the forced properties even in a dry run? (so you can verify or otherwise indicate they would be set?)

ebullient commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/ImagePush.java:
```diff
@@ -31,6 +32,9 @@ public List<Dependency> forcedDependencies() {
     public Map<String, String> forcedProperties() {
+        if (dryRun) {
+            return Collections.emptyMap();
+        }
```
Similar here. Dry run should allow you to inspect behavior.. you'd want to be able to debug or verify what the forced property list is.

Picture future you with a user that this isn't working for..

ebullient commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/ImageTask.java:
```diff
@@ -1,16 +1,29 @@
 
+    boolean dryRun;
+
+    @Option(option = "dry-run", description = "The container image extension to use for building the image (e.g. docker, jib, buildpack, openshift).")
```
Description is incorrect. Steal the text from the CLI option?

https://github.com/quarkusio/quarkus/blob/main/devtools/cli/src/main/java/io/quarkus/cli/common/RunModeOption.java

ebullient commented on devtools/maven/src/main/java/io/quarkus/maven/AbstractImageMojo.java:
```diff
@@ -26,12 +27,22 @@ public enum Builder {
 
+    @Override
+    public void execute() throws MojoExecutionException, MojoFailureException {
+        if (!dryRun) {
```
If it is a dry run.. you should output the properties that could/would have been set. This is your chance to understand what would have happened...

ebullient commented on docs/src/main/asciidoc/cli-tooling.adoc:
```diff
@@ -599,7 +599,5 @@ The `image push` command is similar to `image build`, and surfaces some basic op
 [source, shell]
 ----
-quarkus image push --registry=<<image registry>> --registry-username=<<registry username>> --registry-passowrd=<<registry password>>
+quarkus image push --registry=<<image registry>> --registry-username=<<registry username>> --registry-password-stdin
```
`<<...>>` syntax is used as an anchor.. is that's what intended? 

```suggestion
quarkus image push --registry=<image registry> --registry-username=<registry username> --registry-password-stdin
```

iocanel committed:
Update docs/src/main/asciidoc/cli-tooling.adoc

Co-authored-by: Erin Schnabel <ebullientworks@gmail.com>

iocanel commented:
Dry mode is now supported for all 3 modes: cli, maven mojos & gradle task.

For cli commands run with dry mode, print on screen the underlying maven or gradle command.
For maven mojos run with dry mode print all container image build properties
On gradle dry run seems to run on a lower level and I am not sure if it's possible to print anything as the task is literaly skipped.

iocanel committed:
chore: reorder cli subcommands (bump image)

iocanel commented:
@aloubyansky wanna have a second look before this goes in?

aloubyansky commented:
Introducing not working forced dependencies API in Gradle is bothering me. I'd probably not include that at this point.

iocanel commented:
> Introducing not working forced dependencies API in Gradle is bothering me. I'd probably not include that at this point.

The pull request **does not** `introduce` the forced dependencies api. The api is already there. :smile: 

How would you like me to proceed? Would like to refactor the gradle bits, so that they don't implement `forcedDependencies` at all?

aloubyansky commented:
You did add `forcedDependencies()` and `forcedProperties()` to the Gradle build command. They aren't really working. That's what I mean. While we do want to support this use-case, I am not sure this is the way we are going to support it. We'll need more research on that. WDYT @glefloch?

iocanel committed:
fix: remove obsolete imports
